File: go/reasoning/adapter.go
package reasoning

import (
	"bytes"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"errors"
	"fmt"
	"regexp"
	"strings"
)

// Sentinel errors returned by the reasoning adapter. Every one represents a
// fail-closed outcome: malformed or unbounded input never produces a partial
// tree or partially-stripped content.
var (
	// ErrMalformedInput is returned when the raw input has no usable shape.
	ErrMalformedInput = errors.New("reasoning: malformed input")
	// ErrMalformedJSON is returned when input that looks like JSON cannot be
	// decoded.
	ErrMalformedJSON = errors.New("reasoning: malformed JSON")
	// ErrMalformedReasoningField is returned when a reasoning field is not a
	// string (e.g. reasoning_content is an object or array).
	ErrMalformedReasoningField = errors.New("reasoning: reasoning field is not a string")
	// ErrMalformedContent is returned when a structured content field is
	// neither a string nor null.
	ErrMalformedContent = errors.New("reasoning: content field is not a string")
	// ErrUnbalancedThinking is returned when a thinking tag is opened but
	// never closed, so the rest of the output cannot be trusted.
	ErrUnbalancedThinking = errors.New("reasoning: unbalanced thinking tag")
	// ErrReasoningTooLarge is returned when the extracted reasoning exceeds
	// the configured byte cap (unbounded input).
	ErrReasoningTooLarge = errors.New("reasoning: reasoning exceeds maximum size")
	// ErrTooManyNodes is returned when the reasoning would produce more nodes
	// than configured.
	ErrTooManyNodes = errors.New("reasoning: too many reasoning nodes")
	// ErrTooDeep is returned when the reasoning would produce a deeper chain
	// than configured.
	ErrTooDeep = errors.New("reasoning: reasoning tree too deep")
)

// analysis is the shared result of inspecting a raw model output: the
// extracted reasoning text, the clean content left after stripping, and the
// input format that produced them.
type analysis struct {
	reasoning string
	clean     string
	format    Format
}

// Adapter is a model-agnostic reasoning adapter. It is safe for concurrent
// use after construction; model field registration is read-only afterwards.
type Adapter struct {
	opts       Options
	modelField map[string]string
}

// knownReasoningFields are the field names recognized as reasoning on a
// structured response, in priority order. reasoning_content is the de-facto
// native field (DeepSeek, Qwen/QwQ, Kimi, GLM); the others cover common
// model-specific names.
var knownReasoningFields = []string{
	"reasoning_content",
	"thinking",
	"thinking_content",
	"reasoning",
}

// recognizedKeys are the keys on a structured response that make it worth
// entering the structured path. A mapping without any of these is treated as
// opaque text and returned unchanged.
var recognizedKeys = map[string]bool{
	"message": true, "content": true,
	"reasoning_content": true, "thinking": true,
	"thinking_content": true, "reasoning": true,
}

// New returns a reasoning Adapter with the given options (defaults when none
// are supplied).
func New(opts ...Options) *Adapter {
	o := DefaultOptions()
	if len(opts) > 0 {
		o = opts[0].normalize()
	} else {
		o = o.normalize()
	}
	return &Adapter{
		opts:       o,
		modelField: make(map[string]string),
	}
}

// RegisterModelField maps a model name to a model-specific reasoning field
// name. Fields registered here are consulted before the known field set, so
// callers can teach the adapter names like Spark's thinking payload field.
func (a *Adapter) RegisterModelField(model, field string) {
	if model != "" && field != "" {
		a.modelField[strings.ToLower(model)] = field
	}
}

// Options returns a copy of the adapter's effective options.
func (a *Adapter) Options() Options {
	return a.opts
}

// Strip removes reasoning from raw and returns the clean content that
// ordinary tool-call parsers should see. It fails closed: malformed or
// unbounded input returns an error instead of partially-stripped output.
func (a *Adapter) Strip(raw any) (string, error) {
	an, err := a.analyze(raw, "")
	if err != nil {
		return "", err
	}
	return an.clean, nil
}

// Extract normalizes reasoning found in raw into a versioned context tree.
// Raw reasoning is never stored on the tree: nodes carry bounded summaries.
// The tree has no nodes (RootID "") when raw contains no reasoning.
func (a *Adapter) Extract(raw any, model, backend string) (*ContextTree, error) {
	an, err := a.analyze(raw, model)
	if err != nil {
		return nil, err
	}

	now := a.opts.Clock().UTC()
	tree := &ContextTree{
		Version:   ContractVersion,
		Format:    an.format,
		Model:     model,
		Backend:   backend,
		CreatedAt: now,
		Nodes:     []Node{},
	}

	if strings.TrimSpace(an.reasoning) == "" {
		return tree, nil
	}

	nodes, err := a.parseNodes(an.reasoning, now)
	if err != nil {
		return nil, err
	}
	tree.Nodes = nodes
	if len(nodes) > 0 {
		tree.RootID = nodes[0].ID
	}
	return tree, nil
}

// Parse is Strip and Extract in one call: it returns the clean content and
// the normalized context tree for raw.
func (a *Adapter) Parse(raw any, model, backend string) (string, *ContextTree, error) {
	an, err := a.analyze(raw, model)
	if err != nil {
		return "", nil, err
	}
	tree, err := a.Extract(raw, model, backend)
	if err != nil {
		return "", nil, err
	}
	return an.clean, tree, nil
}

// analyze inspects raw and splits it into reasoning text and clean content.
func (a *Adapter) analyze(raw any, model string) (*analysis, error) {
	switch v := raw.(type) {
	case string:
		return a.analyzeString(v, model)
	case []byte:
		return a.analyzeBytes(v, model)
	case map[string]any:
		return a.analyzeStructured(v, model)
	default:
		return nil, fmt.Errorf("%w: unsupported raw type %T", ErrMalformedInput, raw)
	}
}

func (a *Adapter) analyzeString(s, model string) (*analysis, error) {
	trimmed := strings.TrimSpace(s)
	if strings.HasPrefix(trimmed, "{") {
		var m map[string]any
		err := json.Unmarshal([]byte(trimmed), &m)
		if err == nil && hasRecognizedKey(m) {
			return a.analyzeStructured(m, model)
		}
		if err != nil {
			return nil, fmt.Errorf("%w: %v", ErrMalformedJSON, err)
		}
	}
	return a.analyzeText(s)
}

func (a *Adapter) analyzeBytes(b []byte, model string) (*analysis, error) {
	trimmed := bytes.TrimSpace(b)
	if len(trimmed) > 0 && (trimmed[0] == '{' || trimmed[0] == '[') {
		var v any
		if err := json.Unmarshal(trimmed, &v); err != nil {
			return nil, fmt.Errorf("%w: %v", ErrMalformedJSON, err)
		}
		if m, ok := v.(map[string]any); ok && hasRecognizedKey(m) {
			return a.analyzeStructured(m, model)
		}
	}
	return a.analyzeText(string(b))
}

// analyzeStructured inspects a decoded JSON object. It descends into a
// "message" key (OpenAI-style) and looks for reasoning fields, then for the
// clean content.
func (a *Adapter) analyzeStructured(m map[string]any, model string) (*analysis, error) {
	if msg, ok := m["message"].(map[string]any); ok {
		return a.analyzeStructured(msg, model)
	}

	// Content first: fail closed on a non-string content value.
	var content string
	if c, ok := m["content"]; ok {
		if c == nil {
			content = ""
		} else if cs, ok := c.(string); ok {
			content = cs
		} else {
			return nil, fmt.Errorf("%w: got %T", ErrMalformedContent, c)
		}
	}

	reasoning, format, err := a.extractReasoningField(m, model)
	if err != nil {
		return nil, err
	}

	// A native reasoning field means the content is already clean, but
	// thinking tags may still be embedded in it.
	if reasoning != "" {
		clean, rerr := stripThinkingTags(content)
		if rerr != nil {
			return nil, rerr
		}
		return &analysis{reasoning: reasoning, clean: clean, format: format}, nil
	}

	// No native reasoning field: fall back to thinking tags in the content.
	clean, tags, rerr := splitThinkingTags(content)
	if rerr != nil {
		return nil, rerr
	}
	if tags != "" {
		return &analysis{reasoning: tags, clean: clean, format: FormatThinking}, nil
	}
	return &analysis{reasoning: "", clean: content, format: FormatNone}, nil
}

// extractReasoningField pulls the reasoning text out of a structured object,
// consulting the registered model-specific field first, then the known field
// set in priority order. It returns the detected format.
func (a *Adapter) extractReasoningField(m map[string]any, model string) (string, Format, error) {
	if model != "" {
		if field := a.modelField[strings.ToLower(model)]; field != "" {
			if v, ok := m[field]; ok && v != nil {
				s, ok := v.(string)
				if !ok {
					return "", "", fmt.Errorf("%w: field %q is %T", ErrMalformedReasoningField, field, v)
				}
				if s != "" {
					return s, FormatModelSpecific, nil
				}
			}
		}
	}
	for _, field := range knownReasoningFields {
		v, ok := m[field]
		if !ok || v == nil {
			continue
		}
		s, ok := v.(string)
		if !ok {
			return "", "", fmt.Errorf("%w: field %q is %T", ErrMalformedReasoningField, field, v)
		}
		if s != "" {
			return s, FormatNative, nil
		}
	}
	return "", FormatNone, nil
}

// analyzeText inspects plain text for generic thinking tags.
func (a *Adapter) analyzeText(s string) (*analysis, error) {
	clean, tags, err := splitThinkingTags(s)
	if err != nil {
		return nil, err
	}
	if tags != "" {
		return &analysis{reasoning: tags, clean: clean, format: FormatThinking}, nil
	}
	return &analysis{reasoning: "", clean: s, format: FormatNone}, nil
}

// hasRecognizedKey reports whether a structured object carries any key that
// makes it a reasoning-shaped response.
func hasRecognizedKey(m map[string]any) bool {
	for k := range m {
		if recognizedKeys[k] {
			return true
		}
	}
	return false
}

// thinkingOpen matches the generic thinking open tags, case-insensitively.
var thinkingOpen = regexp.MustCompile(`(?i)<thinking>|\[THINK\]`)

// thinkingClose matches the generic thinking close tags, case-insensitively.
var thinkingClose = regexp.MustCompile(`(?i)</thinking>|\[/THINK\]`)

// splitThinkingTags extracts embedded thinking blocks from text and returns
// the clean text and the joined reasoning. It fails closed on unbalanced
// tags: an open tag without a matching close is an error, not partial output.
func splitThinkingTags(s string) (clean, reasoning string, err error) {
	if !thinkingOpen.MatchString(s) && !thinkingClose.MatchString(s) {
		return s, "", nil
	}

	var b strings.Builder
	var reason strings.Builder
	rest := s
	for {
		loc := thinkingOpen.FindStringIndex(rest)
		if loc == nil {
			// No more open tags; any remaining close tag is unbalanced.
			if thinkingClose.MatchString(rest) {
				return "", "", ErrUnbalancedThinking
			}
			b.WriteString(rest)
			break
		}
		b.WriteString(rest[:loc[0]])
		inner := rest[loc[1]:]
		closeLoc := thinkingClose.FindStringIndex(inner)
		if closeLoc == nil {
			return "", "", ErrUnbalancedThinking
		}
		body := inner[:closeLoc[0]]
		if reason.Len() > 0 {
			reason.WriteByte('\n')
		}
		reason.WriteString(body)
		rest = inner[closeLoc[1]:]
	}
	return b.String(), strings.TrimSpace(reason.String()), nil
}

// stripThinkingTags removes thinking blocks and returns the remaining text.
func stripThinkingTags(s string) (string, error) {
	clean, _, err := splitThinkingTags(s)
	return clean, err
}

// nodeID derives a stable, deterministic node ID from a node's content and
// position so identical reasoning produces identical trees across ports and
// runs.
func nodeID(kind NodeKind, parentID, summary string) string {
	h := sha256.Sum256([]byte(string(kind) + "\x00" + parentID + "\x00" + summary))
	return hex.EncodeToString(h[:])[:16]
}

File: typescript/src/reasoning/adapter.ts
// Model-agnostic reasoning adapter (mirrors go/reasoning/adapter.go).
//
// The adapter recognizes three shapes of model output:
//   - native reasoning_content fields (DeepSeek/Qwen/Kimi/GLM-style JSON)
//   - generic thinking tags (<thinking>...</thinking>, [THINK]...[/THINK])
//   - model-specific reasoning fields (registered per model name)
//
// Reasoning is stripped before ordinary tool-call parsing, and the extracted
// reasoning text is normalized into a versioned context tree. Raw reasoning is
// never exposed in normal replies or default audits: tree nodes carry only
// bounded summaries. The adapter fails closed on malformed or unbounded input.

import { createHash } from "node:crypto";

import {
  MalformedContentError,
  MalformedInputError,
  MalformedJSONError,
  MalformedReasoningFieldError,
  ReasoningError,
  ReasoningTooLargeError,
  TooDeepError,
  TooManyNodesError,
  UnbalancedThinkingError,
} from "./errors.js";
import { normalizeOptions, ReasoningOptions } from "./options.js";
import { CONTRACT_VERSION, ContextTree, Format, Node, NodeKind, NodeStatus } from "./types.js";

export { ReasoningError };

// Known reasoning field names on a structured response, in priority order.
// reasoning_content is the de-facto native field (DeepSeek, Qwen/QwQ, Kimi,
// GLM); the others cover common model-specific names.
const KNOWN_REASONING_FIELDS = ["reasoning_content", "thinking", "thinking_content", "reasoning"];

// Keys that make a structured response worth entering the structured path. A
// mapping without any of these is treated as opaque text and returned
// unchanged.
const RECOGNIZED_KEYS = new Set([
  "message",
  "content",
  "reasoning_content",
  "thinking",
  "thinking_content",
  "reasoning",
]);

const THINKING_OPEN = /<thinking>|\[THINK\]/i;
const THINKING_CLOSE = /<\/thinking>|\[\/THINK\]/i;

const PARAGRAPH_SEP = /\n[ \t]*\n/;
const LIST_ORNAMENT = /^\s*(?:[-*+•]|\d+[.)])\s+/;
const KIND_PREFIX =
  /^\s*(goal|observation|decision|tool-call|tool_call|tool-result|tool_result|tool|conclusion|blocker)\s*[:=]\s*/i;
const STATUS_MARKER = /\s*\[(blocked|pending|active|superseded|complete)\]\s*$/i;
const EVIDENCE_PATTERNS = [
  /\[E\d+\]/g,
  /\bref\s*[=:]\s*(\S+)/g,
  /\bevidence\s*[=:]\s*(\S+)/g,
];
const TOOL_REF = /\b(?:tool|call)\s*[=:]\s*([a-zA-Z_][a-zA-Z0-9_.-]*)/;
const LEADING_IDENTIFIER = /^([a-zA-Z_][a-zA-Z0-9_.-]*)\s*[:=]/;
const BARE_IDENTIFIER = /^[a-zA-Z_][a-zA-Z0-9_.-]*$/;

const KIND_BY_LABEL: Record<string, NodeKind> = {
  goal: NodeKind.GOAL,
  observation: NodeKind.OBSERVATION,
  decision: NodeKind.DECISION,
  "tool-call": NodeKind.TOOL_CALL,
  tool_call: NodeKind.TOOL_CALL,
  tool: NodeKind.TOOL_CALL,
  "tool-result": NodeKind.TOOL_RESULT,
  tool_result: NodeKind.TOOL_RESULT,
  conclusion: NodeKind.CONCLUSION,
  blocker: NodeKind.BLOCKER,
};

const STATUS_BY_LABEL: Record<string, NodeStatus> = {
  blocked: NodeStatus.BLOCKED,
  pending: NodeStatus.ACTIVE,
  active: NodeStatus.ACTIVE,
  superseded: NodeStatus.SUPERSEDED,
  complete: NodeStatus.COMPLETE,
};

interface Analysis {
  reasoning: string;
  clean: string;
  format: Format;
}

function isRecord(v: unknown): v is Record<string, unknown> {
  return typeof v === "object" && v !== null && !Array.isArray(v);
}

export class ReasoningAdapter {
  private readonly opts: Required<ReasoningOptions>;
  private readonly modelFields: Map<string, string>;

  constructor(options?: ReasoningOptions) {
    this.opts = normalizeOptions(options);
    this.modelFields = new Map();
  }

  get options(): Required<ReasoningOptions> {
    return this.opts;
  }

  /** Map a model name to a model-specific reasoning field name. Fields
   * registered here are consulted before the known field set. */
  registerModelField(model: string, fieldName: string): void {
    if (model && fieldName) {
      this.modelFields.set(model.toLowerCase(), fieldName);
    }
  }

  /** Remove reasoning from raw and return the clean content that ordinary
   * tool-call parsers should see. Fails closed. */
  strip(raw: unknown): string {
    return this.analyze(raw, "").clean;
  }

  /** Normalize reasoning found in raw into a versioned context tree. Raw
   * reasoning is never stored on the tree: nodes carry bounded summaries. */
  extract(raw: unknown, model = "", backend = ""): ContextTree {
    const analysis = this.analyze(raw, model);
    const now = this.opts.clock();

    const tree: ContextTree = {
      version: CONTRACT_VERSION,
      format: analysis.format,
      model,
      backend,
      created_at: now,
      root_id: "",
      nodes: [],
    };

    if (analysis.reasoning.trim() === "") {
      return tree;
    }

    const nodes = this.parseNodes(analysis.reasoning, now);
    tree.nodes = nodes;
    if (nodes.length > 0) {
      tree.root_id = nodes[0].id;
    }
    return tree;
  }

  /** Strip and extract in one call: return the clean content and tree. */
  parse(raw: unknown, model = "", backend = ""): { clean: string; tree: ContextTree } {
    const analysis = this.analyze(raw, model);
    return { clean: analysis.clean, tree: this.extract(raw, model, backend) };
  }

  private analyze(raw: unknown, model: string): Analysis {
    if (typeof raw === "string") {
      return this.analyzeString(raw, model);
    }
    if (raw instanceof Uint8Array) {
      return this.analyzeBytes(raw, model);
    }
    if (isRecord(raw)) {
      return this.analyzeStructured(raw, model);
    }
    throw new MalformedInputError(`unsupported raw type ${typeof raw}`);
  }

  private analyzeString(raw: string, model: string): Analysis {
    const trimmed = raw.trim();
    if (trimmed.startsWith("{")) {
      let parsed: unknown;
      try {
        parsed = JSON.parse(trimmed);
      } catch (e) {
        throw new MalformedJSONError(e instanceof Error ? e.message : String(e));
      }
      if (isRecord(parsed) && this.hasRecognizedKey(parsed)) {
        return this.analyzeStructured(parsed, model);
      }
    }
    return this.analyzeText(raw);
  }

  private analyzeBytes(raw: Uint8Array, model: string): Analysis {
    const trimmed = raw.length > 0 ? new TextDecoder().decode(raw).trim() : "";
    if (trimmed.startsWith("{") || trimmed.startsWith("[")) {
      let parsed: unknown;
      try {
        parsed = JSON.parse(trimmed);
      } catch (e) {
        throw new MalformedJSONError(e instanceof Error ? e.message : String(e));
      }
      if (isRecord(parsed) && this.hasRecognizedKey(parsed)) {
        return this.analyzeStructured(parsed, model);
      }
    }
    return this.analyzeText(trimmed);
  }

  private analyzeStructured(m: Record<string, unknown>, model: string): Analysis {
    const message = m["message"];
    if (isRecord(message)) {
      return this.analyzeStructured(message, model);
    }

    let content = "";
    if ("content" in m) {
      const value = m["content"];
      if (value === null || value === undefined) {
        content = "";
      } else if (typeof value === "string") {
        content = value;
      } else {
        throw new MalformedContentError(`content field is ${typeof value}, not a string`);
      }
    }

    const [reasoning, format] = this.extractReasoningField(m, model);

    if (reasoning !== "") {
      return { reasoning, clean: splitThinkingTags(content).clean, format };
    }

    const result = splitThinkingTags(content);
    if (result.reasoning !== "") {
      return { reasoning: result.reasoning, clean: result.clean, format: Format.THINKING };
    }
    return { reasoning: "", clean: content, format: Format.NONE };
  }

  private extractReasoningField(
    m: Record<string, unknown>,
    model: string
  ): [string, Format] {
    if (model) {
      const fieldName = this.modelFields.get(model.toLowerCase());
      if (fieldName) {
        const value = m[fieldName];
        if (value !== null && value !== undefined) {
          if (typeof value !== "string") {
            throw new MalformedReasoningFieldError(
              `field "${fieldName}" is ${typeof value}, not a string`
            );
          }
          if (value !== "") {
            return [value, Format.MODEL_SPECIFIC];
          }
        }
      }
    }
    for (const fieldName of KNOWN_REASONING_FIELDS) {
      const value = m[fieldName];
      if (value === null || value === undefined) {
        continue;
      }
      if (typeof value !== "string") {
        throw new MalformedReasoningFieldError(
          `field "${fieldName}" is ${typeof value}, not a string`
        );
      }
      if (value !== "") {
        return [value, Format.NATIVE];
      }
    }
    return ["", Format.NONE];
  }

  private analyzeText(raw: string): Analysis {
    const result = splitThinkingTags(raw);
    if (result.reasoning !== "") {
      return { reasoning: result.reasoning, clean: result.clean, format: Format.THINKING };
    }
    return { reasoning: "", clean: raw, format: Format.NONE };
  }

  private hasRecognizedKey(m: Record<string, unknown>): boolean {
    return Object.keys(m).some((k) => RECOGNIZED_KEYS.has(k));
  }

  private parseNodes(reasoning: string, now: Date): Node[] {
    if (Buffer.byteLength(reasoning, "utf8") > this.opts.max_reasoning_bytes) {
      throw new ReasoningTooLargeError();
    }

    const paragraphs = splitParagraphs(reasoning);
    if (paragraphs.length > this.opts.max_nodes) {
      throw new TooManyNodesError();
    }

    const nodes: Node[] = [];
    let parentId = "";
    let depth = 0;

    for (const para of paragraphs) {
      depth += 1;
      if (depth > this.opts.max_depth) {
        throw new TooDeepError();
      }

      let kind = NodeKind.OBSERVATION;
      let summary = para.replace(LIST_ORNAMENT, "");
      const kindMatch = KIND_PREFIX.exec(summary);
      if (kindMatch) {
        kind = KIND_BY_LABEL[kindMatch[1].toLowerCase()] ?? NodeKind.OBSERVATION;
        summary = summary.slice(kindMatch[0].length).trim();
      }

      let status = NodeStatus.COMPLETE;
      const statusMatch = STATUS_MARKER.exec(summary);
      if (statusMatch) {
        status = STATUS_BY_LABEL[statusMatch[1].toLowerCase()] ?? NodeStatus.COMPLETE;
        summary = summary.slice(0, statusMatch.index).trim();
      } else if (kind === NodeKind.BLOCKER) {
        status = NodeStatus.BLOCKED;
      }

      summary = truncateCodePoints(summary, this.opts.max_summary_chars);
      if (summary === "") {
        depth -= 1;
        continue;
      }

      const id = nodeId(kind, parentId, summary);
      nodes.push({
        id,
        parent_id: parentId,
        kind,
        status,
        summary,
        evidence_refs: extractEvidence(summary),
        tool_call_refs: extractToolRefs(kind, summary),
        created_at: now,
        updated_at: now,
      });
      parentId = id;
    }
    return nodes;
  }
}

function splitThinkingTags(text: string): { clean: string; reasoning: string } {
  if (!THINKING_OPEN.test(text) && !THINKING_CLOSE.test(text)) {
    return { clean: text, reasoning: "" };
  }

  const parts: string[] = [];
  const reasons: string[] = [];
  let rest = text;
  for (;;) {
    const open = THINKING_OPEN.exec(rest);
    if (open === null) {
      if (THINKING_CLOSE.test(rest)) {
        throw new UnbalancedThinkingError();
      }
      parts.push(rest);
      break;
    }
    parts.push(rest.slice(0, open.index));
    const inner = rest.slice(open.index + open[0].length);
    const close = THINKING_CLOSE.exec(inner);
    if (close === null) {
      throw new UnbalancedThinkingError();
    }
    reasons.push(inner.slice(0, close.index));
    rest = inner.slice(close.index + close[0].length);
  }
  return { clean: parts.join(""), reasoning: reasons.join("\n").trim() };
}

function splitParagraphs(text: string): string[] {
  return text
    .split(PARAGRAPH_SEP)
    .map((p) => p.trim())
    .filter((p) => p !== "");
}

function extractEvidence(summary: string): string[] {
  const out: string[] = [];
  const seen = new Set<string>();
  for (const pattern of EVIDENCE_PATTERNS) {
    pattern.lastIndex = 0;
    let match: RegExpExecArray | null;
    while ((match = pattern.exec(summary)) !== null) {
      const ref = (match[1] ?? match[0]).replace(/^\[|\]$/g, "");
      if (ref !== "" && !seen.has(ref)) {
        seen.add(ref);
        out.push(ref);
      }
    }
  }
  return out;
}

function extractToolRefs(kind: NodeKind, summary: string): string[] {
  if (kind !== NodeKind.TOOL_CALL && kind !== NodeKind.TOOL_RESULT) {
    return [];
  }
  const leading = LEADING_IDENTIFIER.exec(summary);
  if (leading) {
    return [leading[1]];
  }
  const ref = TOOL_REF.exec(summary);
  if (ref) {
    return [ref[1]];
  }
  if (BARE_IDENTIFIER.test(summary)) {
    return [summary];
  }
  return [];
}

function truncateCodePoints(s: string, n: number): string {
  if (n <= 0) {
    return "";
  }
  const points = Array.from(s);
  if (points.length <= n) {
    return s;
  }
  return points.slice(0, n).join("");
}

function nodeId(kind: NodeKind, parentId: string, summary: string): string {
  return createHash("sha256")
    .update(`${kind}\0${parentId}\0${summary}`)
    .digest("hex")
    .slice(0, 16);
}
